xilem_core/src/views/one_of.rs

```diff
@@ -0,0 +1,310 @@
+    G: ViewElement = <Self as PhantomElementCtx>::PhantomElement,
+    H: ViewElement = <Self as PhantomElementCtx>::PhantomElement,
+    I: ViewElement = <Self as PhantomElementCtx>::PhantomElement,
+>: PhantomElementCtx
+{
+    /// Element wrapper, that holds the current view element variant
+    type OneOfElement: ViewElement;
+
+    /// Casts the view element `elem` to the `OneOf::A` variant.
+    /// `f` needs to be invoked with that inner `ViewElement`
+    fn with_downcast_a(elem: &mut Mut<'_, Self::OneOfElement>, f: impl FnOnce(Mut<'_, A>));
+
+    /// Casts the view element `elem` to the `OneOf::B` variant.
+    /// `f` needs to be invoked with that inner `ViewElement`
+    fn with_downcast_b(elem: &mut Mut<'_, Self::OneOfElement>, f: impl FnOnce(Mut<'_, B>));
+
+    /// Casts the view element `elem` to the `OneOf::C` variant.
+    /// `f` needs to be invoked with that inner `ViewElement`
+    fn with_downcast_c(elem: &mut Mut<'_, Self::OneOfElement>, f: impl FnOnce(Mut<'_, C>));
+
+    /// Casts the view element `elem` to the `OneOf::D` variant.
+    /// `f` needs to be invoked with that inner `ViewElement`
+    fn with_downcast_d(elem: &mut Mut<'_, Self::OneOfElement>, f: impl FnOnce(Mut<'_, D>));
+
+    /// Casts the view element `elem` to the `OneOf::E` variant.
+    /// `f` needs to be invoked with that inner `ViewElement`
+    fn with_downcast_e(elem: &mut Mut<'_, Self::OneOfElement>, f: impl FnOnce(Mut<'_, E>));
+
+    /// Casts the view element `elem` to the `OneOf::F` variant.
+    /// `f` needs to be invoked with that inner `ViewElement`
+    fn with_downcast_f(elem: &mut Mut<'_, Self::OneOfElement>, f: impl FnOnce(Mut<'_, F>));
+
+    /// Casts the view element `elem` to the `OneOf::G` variant.
+    /// `f` needs to be invoked with that inner `ViewElement`
+    fn with_downcast_g(elem: &mut Mut<'_, Self::OneOfElement>, f: impl FnOnce(Mut<'_, G>));
+
+    /// Casts the view element `elem` to the `OneOf::H` variant.
+    /// `f` needs to be invoked with that inner `ViewElement`
+    fn with_downcast_h(elem: &mut Mut<'_, Self::OneOfElement>, f: impl FnOnce(Mut<'_, H>));
+
+    /// Casts the view element `elem` to the `OneOf::I` variant.
+    /// `f` needs to be invoked with that inner `ViewElement`
+    fn with_downcast_i(elem: &mut Mut<'_, Self::OneOfElement>, f: impl FnOnce(Mut<'_, I>));
+
+    /// Creates the wrapping element, this is used in `View::build` to wrap the inner view element variant
+    fn upcast_one_of_element(elem: OneOf<A, B, C, D, E, F, G, H, I>) -> Self::OneOfElement;
+
+    /// When the variant of the inner view element has changed, the wrapping element needs to be updated, this is used in `View::rebuild`
+    fn update_one_of_element_mut(
+        elem_mut: &mut Mut<'_, Self::OneOfElement>,
+        new_elem: OneOf<A, B, C, D, E, F, G, H, I>,
+    );
+}
+
+#[doc(hidden)] // Implementation detail, `OneOfState` is public because of trait visibility rules
+mod hidden {
```

Comment:
Hmm yeah, when you mean with consistent, similar as the other views, this would mean no extra module `one_of` at the crate level, and just explicitely exporting all the `OneOfN` types. I wonder generally, whether we should put all views etc. into the top-level without namespacing modules.

Reply:
? I'm not sure I follow

The specific line I'm referring to is the `hidden` module

None of the other "logically hidden" state types use this module technique, and we should be consistent with that

Reply:
Yes, I know, but I think otherwise we need some kind of different workaround like the one before, to hide the state type in the public API. The problem doesn't arise in the other modules, because all the types that should actually be exposed are explicitly `pub use`d

Reply:
Oh, my mistake. I hadn't realised that the other types aren't actually exported.

I was working under the assumption that `unreachable_pub` required that the item in question be nameable, instead of just reachable. I think this is *fine* then, although I really wish there were a lint for paths to be "nameable"

Reply:
There is a separate lint for unnameable types as of 1.79.0: https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#unnameable-types